src/main/solution_space/path_finder/path_finder_v_3.py

```diff
@@ -41,63 +51,58 @@ def find_next_anon_tree(self, user_anon_tree: AnonTree, user_canon_tree: ast.AST
-        """
-
         graph_vertex = self._graph.find_vertex(user_canon_tree)
-        # Todo: find a better way
         if graph_vertex:
             graph_anon_tree = graph_vertex.serialized_code.find_anon_tree(user_anon_tree.tree)
             if graph_anon_tree:
-                return self.__find_closest_tree_from_equal_tree(user_anon_tree, graph_anon_tree)
+                next_anon_trees = [AnonTree.get_item_by_id(id) for id in graph_anon_tree.next_anon_trees_ids]
+                return self.__choose_best_anon_tree(user_anon_tree, next_anon_trees)
+        return None
 
-        canon_nodes_number = get_nodes_number_in_ast(user_canon_tree)
-        anon_nodes_number = user_anon_tree.nodes_number
+    @staticmethod
+    def __get_nodes_number_dict(nodes_number: List[Any], filter_foo: Callable = lambda x: x) -> Dict[int, list]:
+        return {k: list(v) for k, v in groupby(nodes_number, filter_foo)}
+
+    # Note: we have to remove the 'user_code' from the set
+    def __find_closest_tree(self, user_anon_tree: AnonTree, user_canon_nodes_number: int,
+                            user_anon_tree_nodes_number: int,
+                            canon_trees_nodes_number: Dict[int, list]) -> Optional[AnonTree]:
+        """
+        1. Consider each vertex with similar nodes number as candidate (chose at least TOP_N_CANON candidates)
+        """
 
         # Get vertices ids with canon trees, which have nodes number similar to user canon_nodes_number
-        vertices_ids = self.__get_top_n_candidates(CANON_TOP_N, canon_nodes_number, self._graph.canon_trees_nodes_number)
+        vertices_ids = self.__get_top_n_candidates(CANON_TOP_N, user_canon_nodes_number, canon_trees_nodes_number)
         vertices: List[Vertex] = [Vertex.get_item_by_id(id) for id in vertices_ids]
 
         anon_trees = sum([v.serialized_code.anon_trees for v in vertices], [])
-        anon_nodes_numbers_dict = {k: list(v) for k, v in groupby(anon_trees, lambda a_t: a_t.nodes_number)}
-        anon_candidates = self.__get_top_n_candidates(ANON_TOP_N, anon_nodes_number, anon_nodes_numbers_dict)
-
+        anon_nodes_numbers_dict = self.__class__.__get_nodes_number_dict(anon_trees, lambda a_t: a_t.nodes_number)
+        anon_candidates = self.__get_top_n_candidates(ANON_TOP_N, user_anon_tree_nodes_number, anon_nodes_numbers_dict)
         return self.__choose_best_anon_tree(user_anon_tree, anon_candidates)
 
-    def __find_closest_tree_from_equal_tree(self, user_anon_tree: AnonTree, equal_anon_tree: AnonTree) -> AnonTree:
-        log.info('Found the same anon tree')
-        # Todo: what to do if there is the exact vertex? Find a path to the goal? Now just returning it
-        return equal_anon_tree
+    def __is_most_of_path_is_done(self, closest_tree: AnonTree) -> bool:
+        """
+        1. Use only nodes number info. Use median for goals nodes number
+        """
+        return closest_tree.nodes_number <= \
```

Comment:
Shouldn't we use user_anon_tree here instead of closesr_tree? Also, let's call this method '__is_close_to_goals'  instead of '__is_most_of_path_is_done' because now we don't know our path (before we knew that our path is the path to the closest goal, but now we don't have the closest goal)